keps/sig-api-machinery/2608-crd-field-level-versioning/README.md

```diff
@@ -0,0 +1,816 @@
+a similar manner. This KEP aims to provide a mechanism that allows feature-gate-like
+functionality for Custom Resources as well.
+
+## Motivation
+
+Currently, it is not possible to have fields of a CR exist in different stages such
+as alpha, beta, stable or deprecated, the CR itself can be in a version that is defined
+by its Custom Resource Definition (CRD). If a field of a CR were to be marked as deprecated,
+the author of the CRD would have to create a new version of the CRD, mark the current one as
+deprecated, optionally issue a deprecation warning, introduce a new CRD version with the
+deprecated field not present and add conversion logic to convert between the now deprecated
+version and the new version. Please note that these steps are done at the level of an entire
+version, if the deprecated field is not updated/touched after being created, users of this CRD
+would still get deprecation warnings asking them to switch to a new version (provided it is made 
+available).  
+
+Field level versioning for CRs will introduce the ability to have individual fields of a
+CR in different stages and aims to try and simplify the above mentioned process and provide
+a UX similar to built-in types.
+
+### Goals
+
+- Introduce custom feature gates that can be embedded as part of the CRD spec.
+- Define the lifecycle of these custom feature gates and specify what the behaviour
+  of the gate would be at different stages in its lifecycle.
+- Allow issuing custom deprecation warnings for deprecated fields.  
+- Extend the CRD validation chain to include checks for proper usage of these
+  custom feature gates.
+
+### Non-Goals
+
+- Provide a generic mechanism for dynamic feature gate functionality.
+
+## Proposal
+
+This KEP proposes to extend the CRD spec to include feature gate information
+for fields of Custom Resources (CRs):
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: crontabs.stable.example.com
+spec:
+  group: stable.example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                cronSpec:
+                  type: string
+                image:
+                  type: string
+                replicas:
+                  type: integer
+  scope: Namespaced
+  names:
+    ...
+  customFeatureGates:
+    component: istio # optional
+    featureGates:
+      - name: ReplicasFeatureGate
+        enabled: true  # optional
+        default: false # optional
+        preRelease: alpha
+        fieldPaths:
+          - .spec.replicas
+```
+This KEP, on a high level, proposes the following:
+- Multiple feature gates can be defined and each feature gate can then gate
+  multiple `fieldPaths` - each of which is a JSON path for a field in a Custom
+  Resource (CR) of this CRD.
```

Comment:
this design surprises me. It will make the validation hard. It would be **much** easier to have a gate reference in the property specs inside of the schema.